### PR TITLE
define `cache-digest` header in appendix

### PR DESCRIPTION
**Pull request by kazuho**
#256.

Since I wasn't sure if we can make normative references from an appendix, RFC5234 (ABNF) and RFC4648 (Base64) has been added as informative references.

### CODE REVIEW

File: draft-ietf-httpbis-cache-digest.md [hunk 1]
```diff
@@ -286,6 +287,37 @@ Additionally, User Agents SHOULD NOT send CACHE_DIGEST when in "privacy mode."
 
 --- back
 
+# Encoding the CACHE_DIGEST frame as an HTTP Header
+
+Some HTTP/2 protocol stacks do not provide an interface to inject arbitrary HTTP/2 frames while allowing the application to send additional HTTP request headers.
```
Comment (mnot) [inline draft-ietf-httpbis-cache-digest.md]: Isn't the motivation simply that browsers currently don't send CACHE_DIGEST, so you have to fake it with Fetch?

File: draft-ietf-httpbis-cache-digest.md [hunk 2]
```diff
@@ -286,6 +287,37 @@ Additionally, User Agents SHOULD NOT send CACHE_DIGEST when in "privacy mode."
...
+A Cache-Digest request header is defined as a list construct of cache-digest-entities.
+Each cache-digest-entity corresponds to a CACHE_DIGEST frame.
+
+Digest-Value is encoded using base64url {{RFC4648}}, Section 5.
+Flags that are set are encoded as digest-flags by their names that are compared case-insensitively.
+
+Origin is omitted in the header form.
+The value is implied from the value of the `:authority` pseudo header.
+Client MUST only send Cache-Digest headers containing digests that belong to the origin specified by the HTTP request.
+
```
Comment (mnot) [inline draft-ietf-httpbis-cache-digest.md]: Need to register the header field in IANA Considerations, as per RFC3864.

File: draft-ietf-httpbis-cache-digest.md [hunk 3]
```diff
@@ -286,6 +287,37 @@ Additionally, User Agents SHOULD NOT send CACHE_DIGEST when in "privacy mode."
 
 --- back
 
+# Encoding the CACHE_DIGEST frame as an HTTP Header
+
+Some HTTP/2 protocol stacks do not provide an interface to inject arbitrary HTTP/2 frames while allowing the application to send additional HTTP request headers.
+When using such an implementation, it is sensible to send Cache Digests as HTTP headers, even though doing so consumes more bandwidth when compared to using HTTP/2 frames due to the fact that the digests need to be associated to every HTTP request as opposed to just sending once per connection.
+
+For the sake of interoperability with clients that are constrained to using headers, this appendix defines how a CACHE_DIGEST frame can be encoded as an HTTP header named `Cache-Digest`.
+
```
Comment (mnot) [inline draft-ietf-httpbis-cache-digest.md]: It would be good to mention other limitations -- e.g., since the header is end-to-end, there might be confusion about what cache generated the header field.

File: draft-ietf-httpbis-cache-digest.md [hunk 4]
```diff
@@ -286,6 +287,37 @@ Additionally, User Agents SHOULD NOT send CACHE_DIGEST when in "privacy mode."
 
 --- back
 
+# Encoding the CACHE_DIGEST frame as an HTTP Header
+
+Some HTTP/2 protocol stacks do not provide an interface to inject arbitrary HTTP/2 frames while allowing the application to send additional HTTP request headers.
+When using such an implementation, it is sensible to send Cache Digests as HTTP headers, even though doing so consumes more bandwidth when compared to using HTTP/2 frames due to the fact that the digests need to be associated to every HTTP request as opposed to just sending once per connection.
```
Comment (mnot) [inline draft-ietf-httpbis-cache-digest.md]: It might be worth spelling out why they need to be sent on every request (since different requests might be routed differently; there is no connection affinity).

File: draft-ietf-httpbis-cache-digest.md (hunk 1 above)
Comment (kazuho) [inline draft-ietf-httpbis-cache-digest.md]: I have adjusted the text so that it discusses specifically about Service Workers.

It is Service Workers (that provides a Cache _and_ the Fetch API) that we use to emulate Cache-Digests.

**Comment by kazuho** [comment]
Thank you for your review. I believe that I have addressed all four issues in the commit above.

File: draft-ietf-httpbis-cache-digest.md [hunk 5]
```diff
@@ -263,8 +276,8 @@ we can determine whether there is a match in the digest using the following algo
 
 # IANA Considerations
 
-This draft currently has no requirements for IANA. If the CACHE_DIGEST frame is standardised, it
-will need to be assigned a frame type.
+This draft currently has no requirements for IANA.
+If the specification is standardised, the CACHE_DIGEST frame will need to be assigned a frame type and the Cache-Digest header will need to be registered.
```
Comment (mnot) [inline draft-ietf-httpbis-cache-digest.md]: I'm afraid this isn't adequate; if an RFC -- even an experimental one -- defines a header, it really needs to register it. I can do a separate PR if you like.

**Comment by sebdeckers** [comment]
@kazuho Would it make sense to also describe the cookie-based method in this spec?

**Comment by kazuho** [comment]
While we could do that, I am kind of negative against it.

We need a specification for Cache-digest using HTTP/2 frames or HTTP/1 headers, since it needs to be defined as an interaction between a server and a client (that has the knowledge of what is being cached).

OTOH, we do not need a specification for a Cache-Digest cookie, since it is totally controlled by the server. So it is entirely up to each server implementation to use whatever format it wants. Also, it is not easy to find out the right way to implement a cookie-based Cache-Digest that would work on all web browsers, since the timing the browser consumes the cookies sent from the server depend on the timing of the response.

To summarize, the relatively low necessity and the technical difficulty makes me hesitate in pursuing the idea.
